Await peristed state restoration before concluding UI initialization

ravicious opened the pull request:
In #33690, I made it so that the main process waits for the frontend app to finish initialization before sending messages related to deep links. There was one caveat:

> There's one problem though – if the user has opened some tabs and then closed the app, on next launch we show a modal asking if they want to restore the tabs. This is not considered to be a part of UI initialization process – that is because if you're logged into multiple root clusters ("workspaces"), we ask you about restoring tabs for this workspace only just before switching to it.

This was a problem because we want to be sure that there's no modal waiting for the user to interact with, as deep link handlers might themselves trigger some modals (by for example showing the modal for adding a new cluster to which the deep link points to).

The caveat was the result of the fact that while `initUi` calls `workspacesService.restorePersistedState`, the latter does some sync work and then fires an async method. So `initUi` doesn't wait for the state to be restored before resolving.

Since `restorePersistedState` is called only from that place, I figured we might as well make it wait for `setActiveWorkspace`. This means that the UI won't get fully initialized until the user interacts with the login modal (if the cert has expired) and the modal for restoring previous session (if they had any tabs open before closing the app).

## Changes to DocumentsReopen modal

One side effect of this change is that until `initUi` resolves, we don't show any UI other than modals and notifications triggered from `initUi`. This means that when we show DocumentsReopen modal and ask about restoring the previous session, the user has no idea which workspace they're restoring tabs from.

To solve this, I added "You had x tabs open in cluster-name".

This is how app initialization looks like with changes from this PR:

https://github.com/gravitational/teleport/assets/27113/355eb22b-a5c0-4b29-a8f5-6376858e3957

ravicious committed:
Wait for restoring persisted state in initUi

ravicious committed:
DocumentsReopen: Show number of open tabs and cluster name

ravicious committed:
Move the comment for pluralize

avatus reviewed (approved):
Didn't get a chance to try locally yet but code looks good

gzdunek commented:
I have to say that I'm not a fan of that blank screen showing only a dialog:

<img width="975" alt="image" src="https://github.com/gravitational/teleport/assets/20583051/d08ec009-3785-4d83-9824-491b0ac96a41">

IMO it just doesn't look good and makes the UI feel incomplete. 

However, I think that we can easily fix this. We set `setIsUiReady(true)` after [the entire startup sequence is done](https://github.com/gravitational/teleport/blob/794773c35fd1e6746e77584127120c250aa1e060/web/packages/teleterm/src/ui/AppInitializer.tsx#L34). What if we set it before `await initUi(ctx)`? It makes sense to me, because only `ctx.init()` should block showing the UI.

I remember that we added blocking the UI to prevent a flash of 'Connect a Cluster' screen while the app was ran through Rosetta. From what I see, that workaround is still needed (without it I can see the flash of the 'Connect...' screen in a dev mode), but as I said, we should be able to limit it to `ctx.init()` only.

By doing this, we could also avoid showing the cluster name in DocumentsReopen modal.

ravicious committed:
Log errors in AppInitializer

ravicious committed:
Rename functions which initialize the app

ravicious committed:
Don't wait for startup modals before showing UI

ravicious committed:
Create useLogger hook

ravicious commented:
@gzdunek Good idea. I wouldn't say the UI feels incomplete, I think it's mostly what we're used to. But I do agree that if we can make it better with almost no cost then why not.

I renamed a bunch of files because otherwise we had `AppContext.init`, `AppInitializer` and `initUi` and it was not clear how each one was different from the others. Let me know what you think.

I also reverted the change with DocumentsReopen, though I kind of liked it tbh – it gave a clear indication as to what workspace you're restoring the tabs for. Maybe I should just include this change since we already have the code?

ravicious commented on web/packages/teleterm/src/ui/hooks/useLogger.ts:
```diff
@@ -0,0 +1,28 @@
+    loggerRef.current = new Logger(name);
+  }
+
+  return loggerRef.current;
```
I think returning the contents of a ref like this should be fine, as long as that return value is not used for rendering. I couldn't find any arguments against this in the docs.

ravicious commented on web/packages/teleterm/src/ui/AppInitializer.tsx:
```diff
@@ -14,29 +14,37 @@
+      loggerRef.current.error(error?.message);
       setIsUiReady(true);
-      ctx.notificationsService.notifyError(error?.message);
+      appContext?.notificationsService.notifyError(error?.message);
```
This was changed because technically `appContext.init` can throw before it sets up `notificationsService`.

gzdunek reviewed (approved):
>I renamed a bunch of files because otherwise we had AppContext.init, AppInitializer and initUi and it was not clear how each one was different from the others. Let me know what you think.

Much better with them, thanks!

>I also reverted the change with DocumentsReopen, though I kind of liked it tbh – it gave a clear indication as to what workspace you're restoring the tabs for. Maybe I should just include this change since we already have the code?

Yeah, I think we can include it.
